hoomd/md/pytest/test_gsd.py

```diff
@@ -0,0 +1,370 @@
+        snap.constraints.value[0] = 2.5
+
+        # special pairs
+        snap.pairs.types = ['p1', 'p2']
+        snap.pairs.N = 2
+        snap.pairs.typeid[:] = [0, 1]
+        snap.pairs.group[0] = [0, 1]
+        snap.pairs.group[1] = [2, 3]
+
+    return snap
+
+
+def lj_integrator():
+    integrator = hoomd.md.Integrator(dt=0.005)
+    lj = hoomd.md.pair.LJ(nlist=hoomd.md.nlist.Cell(), r_cut=2.5)
+    lj.params.default = {'sigma': 1, 'epsilon': 1}
+    integrator.forces.append(lj)
+    langevin = hoomd.md.methods.Langevin(hoomd.filter.All(), kT=1)
+    integrator.methods.append(langevin)
+
+    return integrator
+
+
+@pytest.fixture(scope='function')
+def create_md_sim(simulation_factory, device, hoomd_snapshot):
+    sim = simulation_factory(hoomd_snapshot)
+    sim.operations.integrator = lj_integrator()
+
+    return sim
+
+@pytest.mark.serial
+def test_write(simulation_factory, hoomd_snapshot, tmp_path):
+    filename = tmp_path / "temporary_test_file.gsd"
+    sim = simulation_factory(hoomd_snapshot)
+    hoomd.write.GSD.write(state=sim.state, mode='wb',
+                          filename=str(filename))
+    if hoomd_snapshot.exists:
+        with gsd.hoomd.open(name=filename, mode='rb') as traj:
+            assert len(traj) == 1
+            assert_equivalent_snapshots(traj[0], hoomd_snapshot)
+
+@pytest.mark.serial
+def test_write_gsd_trigger(create_md_sim, tmp_path):
+
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = create_md_sim
+    gsd_trigger = hoomd.trigger.Periodic(period=10, phase=5)
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=gsd_trigger,
+                                 mode='wb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    sim.run(30)
+
+    if sim.device.communicator.rank == 0:
+        with gsd.hoomd.open(name=filename, mode='rb') as traj:
+            assert [frame.configuration.step for frame in traj] == [5, 15, 25]
+
+@pytest.mark.serial
+def test_write_gsd_mode(create_md_sim, hoomd_snapshot,
+                        tmp_path, simulation_factory):
+
+    filename = tmp_path / "temporary_test_file.gsd"
+
+    sim = create_md_sim
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 mode='wb', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    # run 5 steps and create a gsd file for testing mode=ab
+    snapshot_list = []
+    for _ in range(5):
+        sim.run(1)
+        snap = sim.state.snapshot
+        if snap.exists:
+            snapshot_list.append(snap)
+
+    # test mode=ab
+    sim.operations.writers.clear()
+
+    gsd_writer = hoomd.write.GSD(filename=filename,
+                                 trigger=hoomd.trigger.Periodic(1),
+                                 mode='ab', dynamic=['momentum'])
+    sim.operations.writers.append(gsd_writer)
+
+    snap_list = []
```

Comment:
This method has both `snap_list` and `snapshot_list`. It does make the code confusing to read and understand.

Reply:
Thank you for the suggestion!